mbf_costmap_nav/include/mbf_costmap_nav/search_helper.h

```diff
@@ -0,0 +1,191 @@
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+#ifndef SEARCH_HELPER_H_
+#define SEARCH_HELPER_H_
+
+// mbf
+#include "mbf_costmap_nav/footprint_helper.h"
+#include "mbf_costmap_nav/search_helper_viz.h"
+
+// std
+#include <optional>
+
+// ros
+#include <geometry_msgs/Point.h>
+#include <geometry_msgs/Pose2D.h>
+#include <costmap_2d/costmap_2d_ros.h>
+#include <std_msgs/ColorRGBA.h>
+#include <visualization_msgs/MarkerArray.h>
+
+namespace mbf_costmap_nav
+{
+
+struct SearchConfig
+{
+  double angle_increment;
+  double angle_tolerance;
+  double linear_tolerance;
+  bool use_padded_fp;
+  double safety_dist;
+  geometry_msgs::Pose2D goal;
+};
+
+struct SearchState
+{
+  unsigned char cost;
+  std::uint8_t state{ State::UNSET };
+
+  enum State
+  {
+    FREE = 0,
+    INSCRIBED = 1,
+    LETHAL = 2,
+    UNKNOWN = 3,
+    OUTSIDE = 4,
+    UNSET = 5
+  };
+};
+
+struct SearchSolution
+{
+  geometry_msgs::Pose2D pose;
+  SearchState search_state;
+};
+
+class EuclideanCompare
+{
+private:
+  Cell start_;
+  int start_x_;
+  int start_y_;
+
+public:
+  EuclideanCompare(const Cell& start) : start_(start)
+  {
+    start_x_ = static_cast<int>(start.x);
+    start_y_ = static_cast<int>(start.y);
+  }
+  bool operator()(const Cell& a, const Cell& b)
+  {
+    int ax = static_cast<int>(a.x);
+    int bx = static_cast<int>(b.x);
+    int ay = static_cast<int>(a.y);
+    int by = static_cast<int>(b.y);
+    return std::hypot(ax - start_x_, ay - start_y_) > std::hypot(bx - start_x_, by - start_y_);
+  }
+};
+
+/**
+ * @brief It performs a search on the costmap from start to given distance until it finds a cell where we can place the
+ * robot without colliding with obstacles.
+ * It is a Breadth-depth search performed in a spiral pattern (neighboring cells added in a priority queue), starting
+ * from the start (given) cell. And the cells are ordered by euclidean distance from the start cell (by default but one
+ * can add a new strategy).
+ * The search stops when it finds a cell where the robot can be placed without colliding with obstacles (footprint +
+ * padding).
+ * For each cell, we test multiple angles.
+ * The search is performed on the costmap given in the constructor.
+ */
+class SearchHelper
```

Comment:
Can the members be protected? Making the search method virtual? Maybe later there are other implementations?

Reply:
Why the mix of costmap_2d::Costmap2D* and costmap_2d::Costmap2DROS* in the interface of this class?

Reply:
> Why the mix of costmap_2d::Costmap2D* and costmap_2d::Costmap2DROS* in the interface of this class?

Since the methods are static, the idea is to require the minimum argument for the function to be able to run. 
We have methods that require the `costmap2DROS` wrapper, which is an extension with some extra functionalities; but also methods that can work with the pure `costmap2D` (ROS independent), which is a lighter version.

If they were not static, I would store `costmap2DROS` and use it in all methods, but I prefer to make them static since they may be useful outside; and they are independent functionalities